File: backend/ai/llm_client.py
"""
AI client — placeholder implementations.

Every function here is marked TODO. To integrate a real LLM (e.g. HuggingFace
Inference API, a local Ollama endpoint, or OpenAI-compatible API):
  1. Load the relevant prompt from ai/prompts.py.
  2. Call the model API (async HTTP with httpx is recommended).
  3. Parse the JSON response the model returns.
  4. Replace the mock return value below with the parsed result.

The function signatures and return types must stay the same.
"""

import json
import random
from game.models import ActionResponse, DiplomacyResponse, ResearchResponse, Resource
from ollama import chat
from ai.prompts import GENERATE_RESOURCES_PROMPT, RESOLVE_ATTACK_PROMPT,RESOLVE_DIPLOMACY_PROMPT, RESOLVE_RESEARCH_PROMPT





#TODO: Remove when done testing
from game.map_data import COUNTRIES
print('determining research outcome for Testland...')
# Define a system prompt
system_prompt = "You are a world domination strategy game assistant. Determine outcome of a item or technology research."
# Chat with a system prompt
# # TODO: add separate handling for technology, resource, and item research, outcomes.
response = chat('smollm2', 
                messages=[
                    {'role': 'system', 'content': system_prompt},
                    {'role': 'user', 'content': RESOLVE_RESEARCH_PROMPT.format(player_name="Test", item="zweihander", resources_used=['iron'])}
                ])

result = response.message.content
print('llm generated research outcome', result)











# ---------------------------------------------------------------------------
# Resource generation
# ---------------------------------------------------------------------------

async def generate_country_resources(countries: dict) -> dict[str, dict]:
    print('generating resources for', countries, '...')
    # Define a system prompt
    system_prompt = "You are a world domination strategy game assistant. Generate resources for each country."
    # Chat with a system prompt
    response = chat('smollm2', 
                    messages=[
                        {'role': 'system', 'content': system_prompt},
                        {'role': 'user', 'content': GENERATE_RESOURCES_PROMPT.format(countries=countries)}
                    ])
    
    result = response.message.content
    print('llm generated resources', result)
    try:
        return json.loads(result)
    except json.JSONDecodeError as e:
        raise ValueError("LLM response is not valid JSON")


# ---------------------------------------------------------------------------
# Attack event outcome determination
# ---------------------------------------------------------------------------

async def resolve_attack(
    attacker_name: str,
    attacker_army: int,
    target_country: str,
    target_army: int,
    target_morale: int,
    attack_type: str,
    items_used: list[dict],
    story_context: str,
) -> ActionResponse:
    
    print('determining attack outcome for', attacker_name, 'and', target_country, '...')
    # Define a system prompt
    system_prompt = "You are a world domination strategy game assistant. Determine outcome of an attack."
    # Chat with a system prompt
    response = chat('smollm2', 
                    messages=[
                        {'role': 'system', 'content': system_prompt},
                        {'role': 'user', 'content': RESOLVE_ATTACK_PROMPT.format(attacker_name=attacker_name, attacker_army=attacker_army, target_country=target_country, target_army=target_army, target_morale=target_morale, attack_type=attack_type, items_used=items_used, story_context=story_context)}
                    ])

    result = response.message.content
    print('llm generated attack outcome', result)
    try:
        outcome = json.loads(result)
        action_response = ActionResponse(
            success = outcome.get("success", False),
            new_attacker_army = outcome.get("new_attacker_army_strength", attacker_army),
            new_enemy_army = outcome.get("new_defender_army_strength", target_army),
            story = outcome.get("story", ""),
            found_items = [Resource(**item) for item in outcome.get("items", [])
            ]
        )
    except json.JSONDecodeError as e:
        raise ValueError("LLM response is not valid JSON")
    
    return action_response
    

# ---------------------------------------------------------------------------
# Diplomatic alliance outcome determination
# ---------------------------------------------------------------------------

async def resolve_diplomacy(
    player_name: str,
    target_country: str,
    alliance_type: str,
    resources_used: list[dict],
    story_context: str,
) -> DiplomacyResponse:
    
    print('determining diplomatic alliance outcome for', player_name, 'and', target_country, '...')
    # Define a system prompt
    system_prompt = "You are a world domination strategy game assistant. Determine outcome of a diplomatic alliance."
    # Chat with a system prompt
    response = chat('smollm2', 
                    messages=[
                        {'role': 'system', 'content': system_prompt},
                        {'role': 'user', 'content': RESOLVE_DIPLOMACY_PROMPT.format(player_name=player_name, target_country=target_country, alliance_type=alliance_type, resources_used=resources_used, story_context=story_context)}
                    ])

    result = response.message.content
    print('llm generated diplomatic alliance outcome', result)
    try:
        outcome = json.loads(result)
        diplomacy_response = DiplomacyResponse(
            success = outcome.get("success", False),
            story = outcome.get("story", ""),
            items = [Resource(**item) for item in outcome.get("items", [])
            ]
        )
    except json.JSONDecodeError as e:
        raise ValueError("LLM response is not valid JSON")
    
    return diplomacy_response



# ---------------------------------------------------------------------------
# Research resolution
# ---------------------------------------------------------------------------

async def resolve_research(
    player_name: str,
    item: str,
    resources_used: list[str],
) -> ResearchResponse:
    print('determining research outcome for', player_name, 'and', item, '...')
    # Define a system prompt
    system_prompt = "You are a world domination strategy game assistant. Determine outcome of a research."
    # Chat with a system prompt
    response = chat('smollm2', 
                    messages=[
                        {'role': 'system', 'content': system_prompt},
                        {'role': 'user', 'content': RESOLVE_RESEARCH_PROMPT.format(player_name=player_name, item=item, resources_used=resources_used)}
                    ])

    result = response.message.content
    print('llm generated research outcome', result)
    try:
        outcome = json.loads(result)
        research_response = ResearchResponse(
            success = outcome.get("success", False),
            story = outcome.get("story", ""),
            researched_item = Resource(**outcome.get("researched_item", {})) if outcome.get("researched_item") else {}
        )
    except json.JSONDecodeError as e:
        raise ValueError("LLM response is not valid JSON")
    
    return research_response

File: backend/src/ai/llmClient.ts
import ollama from "ollama";
import {
  generateResourcesPrompt,
  resolveAttackPrompt,
  resolveDiplomacyPrompt,
  resolveResearchPrompt,
} from "./prompts";
import {
  ActionResponse,
  Country,
  DiplomacyResponse,
  Item,
  ResearchResponse,
  Resource,
} from "../game/models";

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

const MODEL = "smollm2";

async function llmRequest(
  systemPrompt: string,
  userPrompt: string,
): Promise<string> {
  const response = await ollama.chat({
    model: MODEL,
    messages: [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ],
  });
  return response.message.content;
}

function parseJson<T>(raw: string): T {
  try {
    return JSON.parse(raw) as T;
  } catch {
    throw new Error("LLM response is not valid JSON");
  }
}

// ---------------------------------------------------------------------------
// Resource generation
// ---------------------------------------------------------------------------

export async function generateCountryResources(
  countries: Record<string, unknown>,
): Promise<
  Record<string, { favoriteResource: string; hatedResource: string }>
> {
  console.log("generating resources for", Object.keys(countries), "...");
  const raw = await llmRequest(
    "You are a world domination strategy game assistant. Generate resources for each country.",
    generateResourcesPrompt(Object.keys(countries)),
  );
  console.log("llm generated resources", raw);
  return parseJson(raw);
}

// ---------------------------------------------------------------------------
// Attack outcome
// ---------------------------------------------------------------------------

export async function resolveAttack(
  attackerEmpire: string[],
  attackerArmyStrength: number,
  attackerMorale: number,
  targetCountry: Country,
  itemsUsed: Resource[],
  storyContext: string,
): Promise<ActionResponse> {
  console.log(
    "determining attack outcome for empire",
    attackerEmpire,
    "and",
    targetCountry.name,
    "...",
  );
  const raw = await llmRequest(
    "You are a world domination strategy game assistant. Determine outcome of an attack.",
    resolveAttackPrompt(
      attackerEmpire,
      attackerArmyStrength,
      attackerMorale,
      targetCountry,
      itemsUsed,
      storyContext,
    ),
  );
  console.log("llm generated attack outcome", raw);
  const outcome = parseJson<Record<string, unknown>>(raw);
  return {
    success: Boolean(outcome.success),
    playerArmyStrength: Number(outcome.newAttackerArmyStrength ?? attackerArmyStrength),
    enemyArmyStrength: Number(outcome.newDefenderArmyStrength ?? targetCountry.armyStrength),
    story: String(outcome.story ?? ""),
    items: (outcome.foundItems as Resource[]) ?? [],
  };
}

// ---------------------------------------------------------------------------
// Diplomacy outcome
// ---------------------------------------------------------------------------

export async function resolveDiplomacy(
  playerEmpireContrieNames: string[],
  targetCountry: Country,
  resourcesUsed: Resource[],
): Promise<DiplomacyResponse> {
  console.log(
    "determining diplomatic alliance outcome for",
    playerEmpireContrieNames.join(", "),
    "and",
    targetCountry.name,
    "...",
  );
  const raw = await llmRequest(
    "You are a world domination strategy game assistant. Determine outcome of a diplomatic alliance.",
    resolveDiplomacyPrompt(
      playerEmpireContrieNames,
      targetCountry,
      resourcesUsed,
    ),
  );
  console.log("llm generated diplomatic alliance outcome", raw);
  const outcome = parseJson<Record<string, unknown>>(raw);
  return {
    success: Boolean(outcome.success),
    story: String(outcome.story ?? ""),
    itemsUsed: (outcome.items as Item[]) ?? [],
    resourceUsed: (outcome.resourceUsed as Resource[]) ?? null,
  };
}

// ---------------------------------------------------------------------------
// Research outcome
// ---------------------------------------------------------------------------

export async function resolveResearch(
  playerName: string,
  item: string,
  resourcesUsed: Resource[],
): Promise<ResearchResponse> {
  console.log(
    "determining research outcome for",
    playerName,
    "and",
    item,
    "...",
  );
  const raw = await llmRequest(
    "You are a world domination strategy game assistant. Determine outcome of a research.",
    resolveResearchPrompt(playerName, item, resourcesUsed),
  );
  console.log("llm generated research outcome", raw);
  const outcome = parseJson<Record<string, unknown>>(raw);
  const researchedItem = outcome.researchedItem as
    | Resource
    | Record<string, never>
    | undefined;
  return {
    success: Boolean(outcome.success),
    story: String(outcome.story ?? ""),
    itemsUsed: (outcome.items as Item[]) ?? [],
    resourceUsed: (outcome.resourceUsed as Resource[]) ?? null,
    researchedItem:
      researchedItem && Object.keys(researchedItem).length > 0
        ? (researchedItem as Resource)
        : null,
  };
}
